### PR TITLE
Adds survey UX (closes #8, closes #59).

### LINKED ISSUE #8: Design user sentiment survey.
We need to finalize the content, wording, and flow of the user sentiment survey.

### LINKED ISSUE #59: Survey UX
Tracking issue for the UX for the survey, which @sevaan is so graciously working on.

### PR DESCRIPTION
**Pull request by chuckharmston**
Video of it in action: https://vid.me/JmnU

I'm going to organize the CSS a bit after opening this up, but wanted to get the PR open as quickly as I could.

### CODE REVIEW

File: src/webextension/survey/components/fields/sentiment.jsx [hunk 1]
```diff
@@ -1,21 +1,58 @@
...
+    );
+  }
+
+  renderLabel(n) {
+    return <label htmlFor={`sentiment-${n}`}>{n}</label>;
+  }
+
+  // This is a bit ugly; it's done this way to ensure that each <label> and
+  // <input> element are siblings, which lets us manage all of the star logic
+  // with CSS alone.
```
Comment (pdehaan) [inline src/webextension/survey/components/fields/sentiment.jsx:30]: +1, this is a bit ~~ugly~~ repetitive. Any way we can use a `for` loop or something?

```js
const stars = [];
for (let id = 5; id >= 1; id--) {
  stars.push(this.renderInput(id));
  stars.push(this.renderLabel(id));
}
```

File: src/webextension/survey/components/fields/sentiment.jsx [hunk 2]
```diff
@@ -1,21 +1,58 @@
...
+  }
+
   render() {
     const { input } = this.props;
     return (
-      <div className="field">
-        <label htmlFor={input.name}>
+      <div className="field field--sentiment">
+        <label>
           How would you rate your experience on sitename?
```
Comment (pdehaan) [inline src/webextension/survey/components/fields/sentiment.jsx:53]: Unrelated: Does "sitename" need to be a variable here?

**Comment by pdehaan** [comment]
For fun (disclaimer: I'm very boring), I tried adding [**eslint-plugin-jsx-a11y**](https://github.com/evcohen/eslint-plugin-jsx-a11y) to our ESLint config, and it found a few potential issues:

### Input:
```json
{
  "env": {
    "browser": true,
    "es6": true,
    "node": true,
    "webextensions": true
  },
  "extends": [
    "eslint:recommended",
    "plugin:react/recommended",
    "plugin:jsx-a11y/recommended",
    "plugin:security/recommended"
  ],
  "globals": {
    "browser": false
  },
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    },
    "sourceType": "module"
  },
  "plugins": [
    "jsx-a11y",
    "react"
  ],
  "rules": {
    "security/detect-object-injection": "off",
    "quotes": ["error", "single"]
  }
}
```

### Output:
```sh
➜  pulse git:(ux) $ npm run lint

> pulse@0.1.0 lint /Users/pdehaan/dev/github/mozilla/pulse
> eslint . --ext .js --ext .jsx

/Users/pdehaan/dev/github/mozilla/pulse/src/webextension/survey/components/fields/sentiment.jsx
  52:9  error  Form controls using a label to identify them must be programmatically associated with the control using htmlFor  jsx-a11y/label-has-for

/Users/pdehaan/dev/github/mozilla/pulse/src/webextension/survey/components/survey.jsx
  33:12  error    Visible, non-interactive elements with click handlers must have at least one keyboard listener  jsx-a11y/click-events-have-key-events
  33:12  warning  Visible, non-interactive elements should not have mouse or keyboard event listeners             jsx-a11y/no-static-element-interactions
  33:12  error    Visible, non-interactive elements with click handlers must have role attribute                  jsx-a11y/onclick-has-role

✖ 4 problems (3 errors, 1 warning)

npm ERR! Darwin 16.4.0
npm ERR! argv "/Users/pdehaan/.nvm/versions/node/v6.9.5/bin/node" "/Users/pdehaan/.nvm/versions/node/v6.9.5/bin/npm" "run" "lint"
npm ERR! node v6.9.5
npm ERR! npm  v3.10.10
```

File: src/webextension/survey/styles.scss [hunk 3]
```diff
@@ -1,6 +1,240 @@
-html {
-  body {
-    background: #FFF;
-    min-width: 320px;
+$font--stack: 'Open Sans', sans-serif;
+$font--default: 13px/1.13 $font--stack;
+
+@font-face {
+  font-family: 'Open Sans';
```
Comment (pdehaan) [inline src/webextension/survey/styles.scss]: Should this be `$font--stack`?

File: src/webextension/survey/styles.scss [hunk 4]
```diff
@@ -1,6 +1,240 @@
...
+}
+
+body {
+  background: #FCFCFC url('../../icons/pulse-64.png') no-repeat 16px 16px / 32px;
+  color: #000;
+  font: $font--default;
+  min-height: 112px;
+  overflow-x: hidden;
+  width: 350px;
+  padding: 16px 16px 16px 64px;
```
Comment (pdehaan) [inline src/webextension/survey/styles.scss]: sass-lint (which we aren't using in this repo) says:

```sh
src/webextension/survey/styles.scss
   30:3   warning  Expected `padding`, found `width`      property-sort-order
   31:3   warning  Expected `width`, found `padding`      property-sort-order
```

---

For future reference (if we want scss-lint), here was my .sass-lint.yml config file:

```yaml
options:
  merge-default-rules: true

files:
  include: 'src/**/*.scss'


rules:
  class-name-format: 0
  extends-before-declarations: 2
  extends-before-mixins: 2
  force-element-nesting: 0
  force-pseudo-nesting: 0
  hex-notation: [2, {style: uppercase}]
  indentation: [2, {size: 2}]
  leading-zero: [2, {include: true}]
  mixins-before-declarations: [2, {exclude: [breakpoint, mq]}]
  nesting-depth: [2, {max-depth: 4}]
  no-color-literals: 0
  no-debug: 1
  no-duplicate-properties: 2
  no-ids: 0
  no-important: 0
  no-qualifying-elements: 0
  no-vendor-prefixes: 0
  no-warn: 1
  placeholder-in-extend: 2
  property-sort-order: [1, {ignore-custom-properties: true}]
```

File: package.json [hunk 5]
```diff
@@ -30,12 +31,14 @@
     "shelljs": "^0.7.6",
     "style-loader": "^0.13.1",
     "urijs": "^1.18.4",
+    "url-loader": "^0.5.7",
```
Comment (pdehaan) [inline package.json:35]: I got this on `npm install`:

```sh
npm WARN url-loader@0.5.7 requires a peer of file-loader@* but none was installed.
```

File: src/webextension/survey/components/fields/sentiment.jsx (hunk 2 above)
Comment (chuckharmston) [reply src/webextension/survey/components/fields/sentiment.jsx:53]: It will be soon. Not quite there yet.

**Comment by chuckharmston** [comment]
> For fun (disclaimer: I'm very boring), I tried adding eslint-plugin-jsx-a11y to our ESLint config, and it found a few potential issues:

https://github.com/mozilla/pulse/issues/92

File: src/webextension/survey/styles.scss (hunk 4 above)
Comment (chuckharmston) [reply src/webextension/survey/styles.scss]: Fixed the issues, then filed https://github.com/mozilla/pulse/issues/93

File: package.json (hunk 5 above)
Comment (chuckharmston) [reply package.json:35]: ✅

**Comment by lmorchard** [comment]
Starting to take a look. But, FWIW, I'd rather wait & assign myself to indicate I'm actually looking rather than get assigned. There's already a request for my review up there, having me assigned might make it look like I've got this when actually I haven't had a chance to get to looking yet

**Comment by lmorchard** [comment]
I'm not super familiar with how Page Actions work, but it seems like I'm not getting the icon button in the URL bar on new empty tabs. 

If I restart or cause the add-on to re-install via file watcher, it sets up icons on all existing tabs. I *do* see a message like `[pulse.webext.lib.util] Created pageAction for 10` in the console when I open a new tab with Cmd-T, but the button doesn't appear. On the other hand, if I Cmd-click a link to spawn a new tab, the icon *does* appear.

File: src/webextension/survey/components/fields/sentiment.jsx [hunk 6]
```diff
@@ -1,21 +1,58 @@
...
+    return <label htmlFor={`sentiment-${n}`}>{n}</label>;
+  }
+
+  // This is a bit ugly; it's done this way to ensure that each <label> and
+  // <input> element are siblings, which lets us manage all of the star logic
+  // with CSS alone.
+  renderStars() {
+    return (
+      <div className="field--star">
+        {this.renderInput(5)}
```
Comment (lmorchard) [inline src/webextension/survey/components/fields/sentiment.jsx:34]: Might be worth throwing a loop in here, but meh. Unrolled loops are fine too.

File: src/webextension/survey/components/survey.jsx [hunk 7]
```diff
@@ -12,6 +13,13 @@ class Survey extends Component {
     return parseInt(val, 10);
   }
 
+  // This is a horrible anti-pattern in React. Unfortunately, we need to force
```
Comment (lmorchard) [inline src/webextension/survey/components/survey.jsx:16]: :nauseated_face: :man_shrugging:

**Review by lmorchard — APPROVED** [review]
Other than the issue I'm seeing with the page action button on new empty pages, this looks :+1: to me

File: src/webextension/survey/components/fields/sentiment.jsx (hunk 6 above)
Comment (chuckharmston) [reply src/webextension/survey/components/fields/sentiment.jsx:34]: They have to be siblings for the CSS to work, but any maps or component class abstraction methods have to return a single element. When [fiber](https://github.com/acdlite/react-fiber-architecture) lands, you'll be able to return arrays of elements. Until then `¯\_(ツ)_/¯`

File: src/webextension/survey/components/survey.jsx (hunk 7 above)
Comment (chuckharmston) [reply src/webextension/survey/components/survey.jsx:16]: Yep. 🤢 is right.

**Pull request merged**